### ISSUE TITLE
Perl and Perl 6 Support

### ISSUE THREAD

**Issue by suic86** [Codewars/codewars-runner-cli#701]
Please complete the following information about the language:

- Name: Perl, Perl 6
- Website: [Perl](perl.org), [Perl6](perl6.org)
- Language Version: Perl 5.28.x or later, [Rakudo Star 2018.10 or later](https://rakudo.org/files/)

The following are optional, but will help us add the language:

- Test Frameworks: 
    - Perl 5: [Test::Unit](https://metacpan.org/pod/Test::Unit), [Test2:Suite](https://metacpan.org/pod/Test2::Suite), [Test::More](https://metacpan.org/pod/Test::More) a there are many others ([wikipedia list](https://en.wikipedia.org/wiki/List_of_unit_testing_frameworks#Perl))
    - Perl 6: [integrated testing framework](https://docs.perl6.org/language/testing), [PSpec i.e. RSpec for Perl 6](https://github.com/supernovus/PSpec)
- How to install: with [cpan command](https://perldoc.perl.org/CPAN.html) or [cpanminus](https://metacpan.org/pod/App::cpanminus)
- How to compile/run: Is described by each framework
- Any comments: Good candidate for test manipulation, regexp and code-golfing and esoteric/domain specific language kata (see [here](https://fosdem.org/2019/schedule/event/perl6newtool/)). It's also a pity that there's Python, Ruby, JS support but not Perl support. :)

---

:+1: reaction _might_ help.

**Comment by Bubbler-4** [comment]
> test manipulation

I believe you meant "text manipulation" :)

**Comment by kazk** [comment]
It might be better to have Perl and Perl6 separate :thinking: 

I have a half finished TAP to Codewars transformer so this shouldn't be difficult.

**Comment by suic86** [comment]
@kazk: IMO they should be separate as they are quite different. This isn't like Python 2.x vs 3.x.
@Bubbler-4: Yes, I meant "text":)

**Comment by kazk** [comment]
@suic86 Yeah, that's what I thought. Can you open a separate Perl6 issue?

**Comment by suic86** [comment]
@kazk: I've created a separate issue for Perl 6.

**Comment by kazk** [comment]
I've played with Test2 and we'll need to write [custom formatter](https://metacpan.org/pod/Test2#Test2::Formatter::). I haven't found any documentation on how to specify the custom formatter though.

The problem is that Test2's TAP output doesn't use [YAML blocks from v13](https://testanything.org/tap-version-13-specification.html) for diagnostic messages. So we need to extract from comments and this is too difficult because it's unstructured. This doesn't work well event-wise either for us because those comments comes after the test is done so it doesn't get associated with correct test.

![image](https://user-images.githubusercontent.com/639336/53295881-ba903080-37b9-11e9-928a-22532233f08b.png)

The only thing left is the formatter, so I can add Perl soon if someone can write `Test2::Formatter::Codewars` and show me how to use it. If not, I'll see if I can figure this out whenever I have the time.

**Comment by suic86** [comment]
@kazk: Have you seen [`Test2::Formatter::TAP`](https://perldoc.pl/Test2::Formatter::TAP)? This could be used be used to get the TAP yaml format for further processing. What do you think?

**Comment by kazk** [comment]
> Have you seen [`Test2::Formatter::TAP`](https://perldoc.pl/Test2::Formatter::TAP)? This could be used be used to get the TAP yaml format for further processing.

Isn't that the default output? I don't think it contains YAML diagnostic blocks. That's TAP, but before v13.

> I've also found [Test::TAPv13](https://metacpan.org/pod/Test::TAPv13).

This allows you to manually add some data to test cases and not attach failure/error details when test fails.

> As many things in Perl are managed via cpan modules, it would make sense to install a reasonable subset of these modules.

We can decide what to include later. I'm already using CPAN to install `Test2` so I'm assuming all we need is the names of packages to install.

**Comment by suic86** [comment]
@kazk: You're right. `Test2::Formatter:TAP` is TAP v12 only and `Test:TAPv13` doesn't provide what you want. I thought that there had been a read-made solution (as it's often the case in Perl). Also, we can compile the CPAN package list later.

**Comment by suic86** [comment]
@kazk: I found two presentations of the main `Test2` developer which might be usefull:

- [Better testing with Test2-Suite](https://www.youtube.com/watch?v=b2xIfBdcqb0)
- [Test2::Harness - Supercharge your test runs](https://www.youtube.com/watch?v=uyqP6Prsq38&t)

As I see the Test2 people don't really like TAP. Using [`yath`](https://metacpan.org/pod/distribution/Test2-Harness/scripts/yath) with a custom [`Test2::Harness::Renderer::Formatter`](https://metacpan.org/pod/Test2::Harness::Renderer::Formatter) (or embracing the existing into CW?) seems to be the right way for Test2. (see [this](https://youtu.be/uyqP6Prsq38?t=479))
